src/main/scala/s_mach/similar/CanSimilar.scala

```diff
@@ -1,7 +1,31 @@
 package s_mach.similar
 
+//import scala.collection.mutable
+
+/**
+ * Utility trait for being able to differentiate instances of A's
+ *
+ * @tparam A type
+ */
 trait CanSimilar[A] {
   def similar(a1: A, a2: A) : Double
-  def cartesianProduct(ma1: IndexedSeq[A], ma2: IndexedSeq[A]) : IndexedSeq[IndexedSeq[Double]] = ??? // default impl here
-  def selfCartesianProduct(ma: IndexedSeq[A]) : IndexedSeq[IndexedSeq[Double]] = ??? // default impl here
-}
+
+
+  def cartesianProduct(ma1: IndexedSeq[A], ma2: IndexedSeq[A]) : IndexedSeq[IndexedSeq[Double]] = {
+    ma1.map(a => ma2.map(similar(a,_)))
+//    val result = mutable.ArraySeq.fill(ma1.length, ma2.length)(1.0)
```

Comment:
All similar ops are commutative and given x, x similar x = 1. This means that we only need to compute the upper half of the matrix and can copy (or pass-thru lookup) the lower half. The diagonal will always be 1.0.